File: Tree_Preorder_Traversal/Tree_Preorder_Traversal.rb
class Node
  def initialize(key)
    @left
    @right
    @val = key
  end
  attr_accessor :left
  attr_accessor :right
  attr_accessor :val
end

def PreOrder(root)
  if root
    print(root.val.to_s + " ")
    PreOrder(root.left)
    PreOrder(root.right)
  end
end

print("PreOrder traversal of tree is ")

root = Node.new(1)
root.left = Node.new(2)
root.right = Node.new(3)
root.left.left = Node.new(4)
root.left.right = Node.new(5)
root.right.left = Node.new(6)
root.right.right = Node.new(7)

PreOrder(root)

# => Output

# PreOrder traversal of tree is 1 2 4 5 3 6 7

File: Tree_Preorder_Traversal/Tree_Preorder_Traversal.ts
/*   Implementation of the preorder traversal of a binary tree in TypeScript language   */

// Declaring the node class
class node {
    right: any;
    value: any;
    left: any;
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

// Recursive Function for Preorder Tree Transversal.
function PreOrder(root) 
{ 
    if (root) {
        console.log(root.value);
        PreOrder(root.left);
        PreOrder(root.right);
    }
} 

// Sample Input
var root = new node(1);
root.left = new node(2);
root.right = new node(3);
root.left.right = new node(4);
root.right.left = new node(8);
root.right.right = new node(7);
root.left.right.left = new node(5);
root.left.right.left.right = new node(6);

// Sample Output
console.log("Preorder traversal of tree formed is:-");
PreOrder(root);


/*  OUTPUT:

Preorder traversal of tree formed is:-
1
2
4
5
6
3
8
7

*/

